### PR TITLE
[SDL-0255] Enhance BodyInformation vehicle data

### LINKED ISSUE #338: [SDL 0255] Enhance BodyInformation vehicle data
Proposal: [Enhance BodyInformation vehicle data](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0255-Enhance-BodyInformation-vehicle-data.md)

> This proposal is to enhance BodyInformation with more params to get access to more door related signals.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/847

### PR DESCRIPTION
**Pull request by ymalovanyi**
Fixes #338

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior

#### Unit Tests
Added unit tests cover [[SDL-0255] Enhance BodyInformation vehicle data](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0255-Enhance-BodyInformation-vehicle-data.md) changes

#### Core Tests
* each of the properties in the struct `DoorStatus` can be received from HMI
* each of the values in the enum `DoorStatusType ` can be received from HMI
* each of the properties in the struct `GateStatus` can be received from HMI
* each of the properties in the struct `RoofStatus` can be received from HMI
* `doorStatuses`, `gateStatuses`, `roofStatuses` properties are present in `BodyInformation` struct
* `GetVehicleData` response message contains `bodyInformation` parameter with `doorStatuses`, `gateStatuses`, `roofStatuses` properties

Core version / branch / commit hash / module tested against: [smartdevicelink/sdl_core#3584](https://github.com/smartdevicelink/sdl_core/pull/3584)
HMI name / version / branch / commit hash / module tested against: [smartdevicelink/sdl_hmi#466](https://github.com/smartdevicelink/sdl_hmi/pull/466)

### Summary
Applied [[SDL-0255] Enhance BodyInformation vehicle data](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0255-Enhance-BodyInformation-vehicle-data.md) changes

##### Breaking Changes
* `N/A`

##### Enhancements
* New structs `DoorStatus`, `GateStatus`, `RoofStatus`
* New enum `DoorStatusType`
* Added required getters/setters in related RPC classes

##### Bug Fixes
* `N/A`

### Tasks Remaining:
* `N/A`

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by ymalovanyi** [comment]
@santhanamk the PR is ready for Ford review.

File: tests/node/rpc/structs/BodyInformationTests.js
Comment (santhanamk) [inline tests/node/rpc/structs/BodyInformationTests.js:33]: @ymalovanyi on line 33 does `Test.JSON_DOORSTATUS_LIST` need to be changed to `Test.GENERAL_DOORSTATUS_LIST`?

File: tests/node/rpc/structs/BodyInformationTests.js
Comment (santhanamk) [inline tests/node/rpc/structs/BodyInformationTests.js:34]: @ymalovanyi on line 34 does `Test.JSON_GATESTATUS_LIST` need to be changed to `Test.GENERAL_GATESTATUS_LIST`?

File: tests/node/rpc/structs/BodyInformationTests.js
Comment (santhanamk) [inline tests/node/rpc/structs/BodyInformationTests.js:35]: @ymalovanyi on line 35 does `Test.JSON_ROOFSTATUS_LIST` need to be changed to `Test.GENERAL_ROOFSTATUS_LIST`?

File: tests/node/rpc/structs/DoorStatusTests.js
Comment (santhanamk) [inline tests/node/rpc/structs/DoorStatusTests.js:18]: @ymalovanyi on line 18 does `Test.JSON_GRID` need to be changed to `Test.GENERAL_GRID`?

File: tests/node/rpc/structs/GateStatusTests.js
Comment (santhanamk) [inline tests/node/rpc/structs/GateStatusTests.js:18]: @ymalovanyi On line 18 does `Test.JSON_GRID` need to be changed to `Test.GENERAL_GRID`?

File: tests/node/rpc/structs/RoofStatusTests.js
Comment (santhanamk) [inline tests/node/rpc/structs/RoofStatusTests.js:19]: @ymalovanyi On line 19 does `Test.JSON_GRID` need to be changed to `Test.GENERAL_GRID`?

File: tests/node/rpc/structs/RoofStatusTests.js
Comment (santhanamk) [inline tests/node/rpc/structs/RoofStatusTests.js:21]: @ymalovanyi On line 21 does `Test.JSON_WINDOW_STATE` need to be changed to `Test.GENERAL_WINDOW_STATE`?

**Review by santhanamk — COMMENTED** [review]
Hi @ymalovanyi I reviewed this PR.  I have a few suggestions on the unit tests.  Please see if anything needs to be changed.

File: tests/Test.js
Comment (ymalovanyi) [reply tests/Test.js:442]: No, because then test will fall, as Grid is a struct.

File: tests/Test.js
Comment (ymalovanyi) [reply tests/Test.js:459]: Same as in above comment.

File: tests/node/rpc/structs/BodyInformationTests.js
Comment (ymalovanyi) [reply tests/node/rpc/structs/BodyInformationTests.js:33]: No, because then test will fall, as DoorStatus is a struct. That is why we need to use list of JSON's.

File: tests/node/rpc/structs/BodyInformationTests.js
Comment (ymalovanyi) [reply tests/node/rpc/structs/BodyInformationTests.js:34]: Same as in comment above, GateStatus is a struct.

File: tests/node/rpc/structs/BodyInformationTests.js
Comment (ymalovanyi) [reply tests/node/rpc/structs/BodyInformationTests.js:35]: Same as in comment above, RoofStatus is a struct.

File: tests/node/rpc/structs/DoorStatusTests.js
Comment (ymalovanyi) [reply tests/node/rpc/structs/DoorStatusTests.js:18]: No, because then test will fail. We're checking for parameters here, and Grid is a struct. That is why we are passing JSON parameters of the struct, but not the whole struct itself.

File: tests/node/rpc/structs/GateStatusTests.js
Comment (ymalovanyi) [reply tests/node/rpc/structs/GateStatusTests.js:18]: Same as in above comment.

File: tests/node/rpc/structs/RoofStatusTests.js
Comment (ymalovanyi) [reply tests/node/rpc/structs/RoofStatusTests.js:19]: Same as in above comment.

File: tests/node/rpc/structs/RoofStatusTests.js
Comment (ymalovanyi) [reply tests/node/rpc/structs/RoofStatusTests.js:21]: No, because WindowState is a struct.

**Review by santhanamk — APPROVED** [review]
@ymalovanyi thank you for the clarification.  I am approving this PR.

**Comment by ymalovanyi** [comment]
@santhanamk, thank you for the approval! @crokita, Ford approved this PR, please review.

**Comment by jordynmackool** [comment]
Hi @ymalovanyi and @santhanamk, since the core PR for this feature is not complete, it seems that this PR would have not been tested fully. To ensure the review process goes smoothly:

1. Please avoid removing sections from the PR template as all the information requested is needed.
2. Please provide the Core branch/PR that this was tested against.

Once this PR is fully tested, tag me and Livio will review. Thank you!

**Comment by vladmu** [comment]
@jordynmackool @crokita the description was fixed to include all sections from the template and the code was tested against the Core and HMI (corresponded links also included in the description). Please review.

**Comment by vladmu** [comment]
@crokita, the PR was updated with the latest `develop`, please don't merge this PR after the approval before this related `rpc_spec` PR https://github.com/smartdevicelink/rpc_spec/pull/297 is not merged.

**Review by renonick87 — APPROVED** [review]
This PR is approved. Waiting on [smartdevicelink/rpc_spec#297](https://github.com/smartdevicelink/rpc_spec/pull/297) to be approved before merging.

**Pull request merged**
